### PR TITLE
mem: add mprotect tests

### PR DESCRIPTION
**Pull request by badochov**
JIRA: RTOS-666

## Description
Add tests to new mprotect syscall.

Demo PR showing that tests pass with kernel and libphoenix changes applied.
https://github.com/phoenix-rtos/phoenix-rtos-project/pull/922
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/475
https://github.com/phoenix-rtos/libphoenix/pull/311
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: mem/Makefile [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
+LOCAL_SRCS := test_mprotect.c
+DEP_LIBS := unity
+
+include $(binary.mk)
+
+NAME := test-mprotect-fault
+LOCAL_SRCS := test_mprotect_fault.c
+DEP_LIBS := unity
+
+include $(binary.mk)
```
Comment (agkaminski) [inline mem/Makefile:25]: No new line at the end of the file

File: mem/test_mprotect.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+
+TEST_TEAR_DOWN(test_mprotect)
+{
+}
+
+
+TEST(test_mprotect, test_mprotect_singlecore)
+{
+	unsigned char *area = mmap(0, page_size * PAGES, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
```
Comment (agkaminski) [inline mem/test_mprotect.c:43]:
```suggestion
	unsigned char *area = mmap(NULL, page_size * PAGES, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
```

File: mem/test_mprotect.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+TEST(test_mprotect, test_mprotect_multicore)
+{
+	pthread_t thread;
+	struct mprotect_data data = {
+		.state = PROTECT,
+		.size = page_size * PAGES,
+		.prot = PROT_NONE,
+	};
+	data.area = mmap(0, page_size * PAGES, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	TEST_ASSERT_NOT_NULL(data.area);
```
Comment (agkaminski) [inline mem/test_mprotect.c:125]: mmap returns `MAP_FAILED` on fail, not NULL

File: mem/test_mprotect.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+TEST_TEAR_DOWN(test_mprotect)
+{
+}
+
+
+TEST(test_mprotect, test_mprotect_singlecore)
+{
+	unsigned char *area = mmap(0, page_size * PAGES, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	TEST_ASSERT_NOT_NULL(area);
```
Comment (agkaminski) [inline mem/test_mprotect.c:44]: mmap returns `MAP_FAILED` on fail, not NULL

File: mem/test_mprotect.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+	do {
+		TEST_ASSERT_EQUAL(0, pthread_cond_wait(&data->cond, &data->mutex));
+	} while (data->state != USE);
+
+	return data->res;
+}
+
+
+TEST(test_mprotect, test_mprotect_multicore)
```
Comment (agkaminski) [inline mem/test_mprotect.c:116]: I am not sure about this test. We don't have any guarantees that threads will be assigned to different cores, also threads seem to be serialized by lock and cond, so aren't run in parallel anyway

File: mem/test_mprotect.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	RUN_TEST_GROUP(test_mprotect);
+}
+
+
+int main(int argc, char *argv[])
+{
+	UnityMain(argc, (const char **)argv, runner);
+
+	return 0;
+}
```
Comment (agkaminski) [inline mem/test_mprotect.c:181]: No new line

File: mem/test_mprotect_fault.c [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+{
+}
+
+
+TEST(test_mprotect_fault, unit)
+{
+	long pageSize = sysconf(_SC_PAGESIZE);
+	long totalSize = 4 * pageSize;
+	volatile unsigned char *area = mmap(NULL, totalSize, PROT_READ | PROT_WRITE, MAP_ANONYMOUS, -1, 0);
+	TEST_ASSERT_NOT_NULL(area);
```
Comment (agkaminski) [inline mem/test_mprotect_fault.c:42]: MAP_FAILED

File: mem/test_mprotect_fault.c [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+	TEST_ASSERT_EQUAL(0, mprotect((void *)area, totalSize, PROT_READ));
+
+	area[0] = 0x42;                   /* Should generate fault. */
+	TEST_ASSERT_EQUAL(0x42, area[0]); /* Here to make sure compiler doesn't optimize away write. */
```
Comment (agkaminski) [inline mem/test_mprotect_fault.c:47]: Hmm, not needed, volatile musn't be optimized-out, also if we managed to protect this memory it cannot assume value of 0x42
```suggestion
	TEST_ASSERT_NOT_EQUAL(0x42, area[0]);
```

File: mem/test_mprotect_fault.c [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	RUN_TEST_GROUP(test_mprotect_fault);
+}
+
+
+int main(int argc, char *argv[])
+{
+	UnityMain(argc, (const char **)argv, runner);
+
+	return 0;
+}
```
Comment (agkaminski) [inline mem/test_mprotect_fault.c:68]: No new line

**Review by damianloew — COMMENTED** [review]
Thanks for adding such simple tests, I think we will extend them in the future to provide better coverage for this function.

File: mem/test_mprotect.c [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+
+TEST_TEAR_DOWN(test_mprotect)
+{
+}
+
+
+TEST(test_mprotect, test_mprotect_singlecore)
+{
+	unsigned char *area = mmap(NULL, page_size * PAGES, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
```
[the remark this answers is not in the document]
Comment (badochov) [reply mem/test_mprotect.c:46]: I wouldn't work then as we require that the protection required by the mrpotect has to be lesser or equal than the one established by mmap

File: mem/test_mprotect.c [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+static long page_size;
+
+
+TEST_GROUP(test_mprotect);
+
+
+TEST_SETUP(test_mprotect)
+{
+	page_size = sysconf(_SC_PAGESIZE);
```
[the remark this answers is not in the document]
Comment (badochov) [reply mem/test_mprotect.c:35]: moved to the body

**Review by damianloew — APPROVED** [review]
LGTM

**Pull request merged**